Feature/dojo guestbook

asv-tungpt opened the pull request:
Tạm thời em vẫn chưa thể chạy được dojo, em sẽ cố gắng fix.

asv-tungpt committed:
Create api

asv-tungpt committed:
Submit api

asv-tungpt committed:
Update file

asv-tungpt committed:
Test api successful

asv-tungpt committed:
Remove unnecessary post method

asv-tungpt committed:
Repair some issues

asv-tungpt committed:
Dojo-guestbook

phucnpm commented on djangoguestbook/settings.py:
```diff
@@ -83,7 +83,7 @@
 	'django.middleware.common.CommonMiddleware',
 	'django.contrib.sessions.middleware.SessionMiddleware',
-	'django.middleware.csrf.CsrfViewMiddleware',
+	# 'django.middleware.csrf.CsrfViewMiddleware',
```
không tắt CRSF, 
để JS giao tiếp dc với Django, hãy chèn X-CSRFToken vào HTTP request header 
https://docs.djangoproject.com/en/dev/ref/csrf/
http://www.lemonsoftware.eu/goodies/python:django-dojo-csrf-token

phucnpm replied:
không comment line 86: `'django.middleware.csrf.CsrfViewMiddleware',`

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/dojo_guestbook.html:
```diff
@@ -0,0 +1,45 @@
+    require([
+        "dojo/parser",
+        "dojo/_base/config",
+        "myGuestbook/views/GuestbookView",
```
root nên là app.js
hãy nạp các Widget / View của em bên trong app.js 
https://github.com/asv-rak/gae-dojo-sample/blob/develop/static/js/myapp/app.js

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,143 @@
+], function(declare, baseFx, lang, dom, domStyle, mouse, on,
+            _WidgetBase, _TemplatedMixin, _WidgetsInTemplateMixin,
+            InlineEditBox, template, GreetingStore){
+    return declare("guestbook.Greeting", [_WidgetBase, _TemplatedMixin, _WidgetsInTemplateMixin], {
```
hãy đặt tên class trùng với tên file JS
`return declare("guestbook.GreetingView", ... )`

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,117 @@
+
+        _showSignGreetingForm: function(){
+            this.signFormWidget = new SignFormWidget({GuestbookViewParent:this});
+            this.signFormWidget.placeAt(this.signFormContainerNode);
```
nếu luôn hiển thị trong GuestbookView thì hãy nhúng vào template dùng data-dojo-type
http://dojotoolkit.org/documentation/tutorials/1.8/declarative/

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,117 @@
+        },
+
+        _showSignGreetingForm: function(){
+            this.signFormWidget = new SignFormWidget({GuestbookViewParent:this});
```
có thể get parent widget bằng cách sau
dijit.getEnclosingWidget(widget.domNode.parentNode);

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,143 @@
+        },
+
+        _onclickSaveBtn: function(){
+            var _guestbookParent = this.GuestbookViewParent;
```
không lạm dụng prefix  `_` trong tên biến
chỉ dùng cho private (helper) method

phucnpm commented on djangoguestbook/static/js/guestbook/views/models/GreetingStore.js:
```diff
@@ -0,0 +1,112 @@
+        deleteGreeting: function(guestbookName, greetingId){
+            this.set('guestbookName', guestbookName);
+
+            var deferred = new Deferred();
```
nếu không có xử lý thêm gì chỗ này thì hãy
`return this.store.remove(greetingId);`
không cần tạo thêm Deferred, tương tự cho những chỗ khác

asv-tungpt committed:
update guestbook dojo

- Show greeting được
- Update được
- Delete và add được, nhưng ko load đc trang, phải F5 mới refresh

phucnpm commented:
sửa lại cách tổ chức file
models nằm ngang cấp với views 

```
-js/guestbook
   |-views
        |-viewA.js
   |-models
        |-modelA.js
   |-tests
        |-views
             |-viewA.js
        |-models
             |-modelA.js
```

phucnpm commented on djangoguestbook/guestbook/templates/guestbook/dojo_guestbook.html:
```diff
@@ -0,0 +1,45 @@
+    require([
+        "dojo/parser",
+        "dojo/_base/config",
+        "myGuestbook/GuestbookView",
```
sửa lại theo cấu trúc sau: 
- trong template hãy load app.js
  - trong app.js mới load GuestbookView

Xem
https://github.com/asv-rak/gae-dojo-sample/blob/develop/static/js/myapp/app.js

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,145 @@
+], function(declare, baseFx, lang, dom, domStyle, mouse, on,
+			_WidgetBase, _TemplatedMixin, _WidgetsInTemplateMixin,
+			InlineEditBox, template, GreetingStore){
+	return declare("guestbook.GreetingView", [_WidgetBase, _TemplatedMixin, _WidgetsInTemplateMixin], {
```
hãy thêm _ViewBaseMixin.js 
https://github.com/asv-rak/gae-dojo-sample/blob/develop/static/js/myapp/views/_ViewBaseMixin.js

GuestbookView & GreetingView hãy kế thừa  _ViewBaseMixin.js

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,118 @@
+	"dojo/on",
+	"dojo/dom",
+	"dojo/dom-construct",
+	"dojo/_base/array",
```
sửa lại thứ tự import module trong JS (kiểm tra lại tất cả các file), xem 
http://rel.aoi-sys.vn/redmine/projects/rakumo/wiki/Coding_style#1-Import-module-rule

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,145 @@
+				// handle button Delete
+				on(this.deleteButtonNode, "click", lang.hitch(this, "_onclickDeleteBtn"))
+				// button save in InLineEditText
+				,on(this.contentNode, "change", lang.hitch(this, "_onclickSaveBtn"))
```
dấu `,` dời lên cuối line 63

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,145 @@
+			var greetingId = this.greetingIdNode.value;
+
+			var contentLength = greetingContent.length;
+			if (contentLength > 0 && contentLength <= 10){
```
dùng dijit.form.ValidationTextBox để xử lý ràng buộc content và hiển thị lỗi

phucnpm replied:
nếu ValidationTextBox đã cho pass thì không cần kiểm tra lại nữa

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,145 @@
+			deleteGreetingDeferred.then(function(results){
+					guestbookParent.reloadListGreeting(guestbookName);
+				},
+				function(err){
```
sửa lại convention, 
xem sample

```
    process.then(function(results){
      dom.byId("output").innerHTML += "<br/>I'm finished, and the result was: " + results;
    }, function(err){
      dom.byId("output").innerHTML += "<br/>I errored out with: " + err;
    }, function(progress){
      dom.byId("output").innerHTML += "<br/>I made some progress: " + progress;
    });
```

phucnpm commented on djangoguestbook/static/js/guestbook/views/GreetingView.js:
```diff
@@ -0,0 +1,145 @@
+				alert("Error: This content is empty or length > 10 chars")
+			}
+		}
+		, setHiddenDeleteNode:function(hidden){
```
trình bày class theo convention của array, không dùng dấu `,` để bắt đầu 1 line, hãy dời lên cuối line:127 (sửa luôn những chỗ khác

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,118 @@
+		},
+
+		_showListGreeting: function(guestbookName){
+			var isAdmin = "false";
```
`"false"` là string, nếu khởi tạo là boolean thì là `false` (không có " " )

phucnpm replied:
không handle string-bool nhập nhằng kiểu này, 
hãy dùng dojoConfig để lưu giá trị is_user_admin (boolean), 

tại `djangoguestbook/guestbook/templates/guestbook/dojo_guestbook.html`

```
dojoConfig{
 ...
 is_user_admin: {% if is_user_admin %}true{% else %}false{% endif %},
}
```

xem cách xài dojoConfig
http://dojotoolkit.org/documentation/tutorials/1.9/dojo_config/

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,118 @@
+				isAdmin = isUserAdminNode.value;
+			}
+
+			var userLogin = "false";
```
đây không phải biến bool, không khởi tạo bằng string này có thể gây nhầm lẫn với bool
sửa lại `var userLogin = "";`
hãy đưa thông tin userLogin vào dojoConfig, tương tự https://github.com/asv-rak/python-training/pull/18/files#r25363355

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,118 @@
+		},
+
+		_removeAllGreeting: function(){
+			this.greetingsContainerNode.innerHTML = "";
```
empty innerHTML là không đúng, các Greeting vẫn tồn tại trong vùng nhớ
hãy dùng destroyRecursive
xem https://dojotoolkit.org/documentation/tutorials/1.9/understanding_widgetbase/

asv-tungpt replied:
có code mẫu cách gọi destroyRecursive, destroyRendering ko anh? Em đã thử nhưng ko được

phucnpm commented on djangoguestbook/static/js/guestbook/views/SignFormWidget.js:
```diff
@@ -0,0 +1,58 @@
+
+		_signNewGreeting: function(guestbookView, guestbookName, greetingContent){
+			var _contentLength = greetingContent.length;
+			if (_contentLength > 0 && _contentLength <= 10){
```
hãy dùng dijit.form.ValidationTextBox để xử lý chỗ này

asv-tungpt replied:
Em đã có xử lí trong template SignFormWidget rồi anh: 
<input type="text" value="" required="true"
            data-dojo-attach-point="contentNode"
            data-dojo-type="dijit/form/ValidationTextBox"
            data-dojo-props="pattern:'.{10}', invalidMessage:'Max of content is 10 chars.'" />
Code trong file js là để ngăn chặn lúc người dùng vẫn bất chấp nhập quá 10 kí tự và nhấn sign button

phucnpm commented on djangoguestbook/static/js/guestbook/views/models/GreetingStore.js:
```diff
@@ -0,0 +1,104 @@
+		},
+
+		// get list greeting
+		getListGreeting: function(guestbookName){
```
làm gọn lại tương tự deleteGreeting, 
sửa luôn những method khác nếu có thể

asv-tungpt committed:
Repair some comment

asv-tungpt committed:
Load app.js và kế thừa _ViewBaseMixin

- GuestbookView kế thừa _ViewBaseMixin thì được.
- GreetingView kế thừa _VIewBaseMixin thì

asv-tungpt committed:
Remove alert

Vẫn chưa thể kế thừa _BaseViewMixin từ

asv-tungpt committed:
Repair _ViewBaseMixin

phucnpm commented on djangoguestbook/static/js/guestbook/models/GreetingStore.js:
```diff
@@ -0,0 +1,95 @@
+
+			var deferred = new Deferred();
+			var contentLength = greetingContent.length;
+			if (contentLength > 0 && contentLength <= 10){
```
store chỉ có nhiêm vụ access API
không kiểm tra dữ liệu trong này

phucnpm commented on djangoguestbook/static/js/guestbook/models/GreetingStore.js:
```diff
@@ -0,0 +1,95 @@
+
+			var deferred = new Deferred();
+			var _contentLength = greetingContent.length;
+			if (_contentLength > 0 && _contentLength <= 10){
```
tương tự, k0 kiểm tra dữ liệu trong này

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,113 @@
+
+		_onclickSwitchBtn: function(){
+			var guestbookNameLength = this.guestbookNameNode.value;
+			if (guestbookNameLength > 0 && guestbookNameLength <= 20){
```
ValidateTextbox cho pass thì không cần test lại

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,113 @@
+			this.guestbookNameNode.set("value", guestbookName);
+		},
+
+		reloadListGreeting:function(guestbookName, greetingID){
```
param greetingID không dùng thì bỏ đi

phucnpm commented on djangoguestbook/static/js/guestbook/views/GuestbookView.js:
```diff
@@ -0,0 +1,113 @@
+	"dojo/dom",
+	"dojo/dom-construct",
+	"dojo/text!./templates/GuestbookView.html",
+	"dijit",
```
sort lại thứ tự import (kiểm tra lại những file khác nữa )
- dijit không xài thì bỏ đi

asv-tungpt committed:
Fix some comment
